File: frontend/src/types/verification.ts
export interface DetectedFaces {
  selfie: number
  document: number
}

export interface VerificationResult {
  matched: boolean
  decision: VerificationDecision
  similarityPercentage: number
  thresholdPercentage: number
  detectedFaces: DetectedFaces
  quality: FaceQualitySummary
}

export type VerificationDecision = 'match' | 'no_match' | 'inconclusive'

export interface FaceQualitySummary {
  selfieDetectionConfidence: number
  documentDetectionConfidence: number
  warnings: QualityWarning[]
}

export type QualityWarning =
  | 'LOW_SELFIE_FACE_CONFIDENCE'
  | 'LOW_DOCUMENT_FACE_CONFIDENCE'
  | 'SELFIE_FACE_TOO_SMALL'
  | 'DOCUMENT_FACE_TOO_SMALL'
  | 'LOW_SELFIE_FACE_SHARPNESS'
  | 'LOW_DOCUMENT_FACE_SHARPNESS'
  | 'POOR_SELFIE_FACE_EXPOSURE'
  | 'POOR_DOCUMENT_FACE_EXPOSURE'

export type VerificationErrorCode =
  | 'INVALID_IMAGE'
  | 'IMAGE_TOO_LARGE'
  | 'UNSUPPORTED_IMAGE_TYPE'
  | 'FACE_NOT_FOUND'
  | 'MULTIPLE_FACES_FOUND'
  | 'FACE_COMPARISON_FAILED'

export interface VerificationError {
  code: VerificationErrorCode
  message: string
  field?: 'selfie' | 'document' | null
}

export type VerificationRequestStatus = 'idle' | 'submitting' | 'success' | 'error'

File: backend/app/schemas/verification.py
from enum import StrEnum

from pydantic import BaseModel, Field

from app.core.errors import ErrorCode


class FaceDetectionSummary(BaseModel):
    selfie: int = Field(ge=0)
    document: int = Field(ge=0)


class VerificationDecision(StrEnum):
    MATCH = "match"
    NO_MATCH = "no_match"
    INCONCLUSIVE = "inconclusive"


class FaceQualitySummary(BaseModel):
    selfie_detection_confidence: float = Field(ge=0, le=100)
    document_detection_confidence: float = Field(ge=0, le=100)
    warnings: list[str]


class VerificationResponse(BaseModel):
    matched: bool
    decision: VerificationDecision
    similarity_percentage: float = Field(ge=0, le=100)
    threshold_percentage: float = Field(ge=0, le=100)
    detected_faces: FaceDetectionSummary
    quality: FaceQualitySummary


class ErrorResponse(BaseModel):
    code: ErrorCode
    message: str
    field: str | None = None
